lib/evmone/symbolic.hpp

```diff
@@ -0,0 +1,254 @@
+struct SetMem;
+
+
+using SymbolicMemoryUpdate =
+    std::variant<SetItem, SetMem, Offset>;
+std::ostream& operator<<(std::ostream& os, const SymbolicMemoryUpdate& i);
+
+
+
+using SymbolicStorage = SymbolicUpdates<SetItem>;
+using SymbolicStoragePtr = SymbolicUpdatesPtr<SetItem>;
+
+
+struct MapComparator
+{
+    bool operator()( const evmc_address& a, const evmc_address& b ) const 
+    {
+        for (size_t i = 0; i < 20; i++)
+        {
+            if(a.bytes[i]<b.bytes[i]) return true;
+        }
+        return false;
+    }
+};
+
+using SymbolicStorageMap = std::map<evmc_address, SymbolicStoragePtr, MapComparator>;
+
+using SymbolicMemory = SymbolicUpdates<SymbolicMemoryUpdate>;
+using SymbolicMemoryPtr = SymbolicUpdatesPtr<SymbolicMemoryUpdate>;
+using PlainMemoryPtr = std::shared_ptr<uint8_t[]>;
+
+using Sload = Load<SymbolicStoragePtr>;
+using Mload = Load<SymbolicMemoryPtr>;
+
+using SymbolicStackItem =
+        std::variant<Pure, Sload, Mload, UnaryOp, BinaryOp, TernaryOp>;
+using SymbolicStackItemPtr = rc_ptr<SymbolicStackItem>;
+
+std::ostream& operator<<(std::ostream& os, const SymbolicStackItem& i);
+
+
+struct Offset {
+    SymbolicStackItemPtr offset;
+    SymbolicStackItemPtr size;
+    friend std::ostream& operator<<(std::ostream&, const Offset&);
+};
+
+struct SetMem
+{
+    size_t index;
+    size_t size;
+    std::variant<SymbolicMemoryPtr, std::shared_ptr<uint8_t[]>> memory;
+    friend std::ostream& operator<<(std::ostream&, const SetMem&);
+};
+
+struct SetItem
+{
+    SymbolicStackItemPtr loc;
+    SymbolicStackItemPtr val;
+    friend std::ostream& operator<<(std::ostream&, const SetItem&);
+};
+
+template<typename T>
+struct SymbolicUpdates 
+{
+    // we shouuld only ever have one evmc_address per SymbolicUpdates list
+    std::variant<T, evmc_address> head;
+    SymbolicUpdatesPtr<T> tail;
+    template<typename U>
+    friend std::ostream& operator<<(std::ostream&, SymbolicUpdatesPtr<U>);
+};
+
+struct Pure
+{
+    uint256 pure;
+};
+
+template<typename T>
+struct Load 
+{
+    SymbolicStackItemPtr addr;
+    T symbolic_store;
+};
+
+enum class UnOp { iszero, not_ };
+std::ostream& operator<< (std::ostream&, const UnOp&);
+
+struct UnaryOp
+{
+    UnOp op;
+    SymbolicStackItemPtr first;
+};
+
+enum class BinOp { add, mul, sub, div, sdiv, mod, smod, exp, signextend, lt, gt, slt, sgt, eq, and_, or_, xor_, byte, shl, shr, sar, keccak256 };
+std::ostream& operator<< (std::ostream&, const BinOp&);
+
+struct BinaryOp
+{
+    BinOp op;
+    SymbolicStackItemPtr first;
+    SymbolicStackItemPtr second;
+};
+
+enum class TernOp { addmod, mulmod };
+std::ostream& operator<< (std::ostream&, const TernOp&);
+
+struct TernaryOp
+{
+    TernOp op;
+    SymbolicStackItemPtr first;
+    SymbolicStackItemPtr second;
+    SymbolicStackItemPtr third;
+};
+
+template <bool Symbolic>
+struct StackItem;
+
+template <>
+struct StackItem<false> {
+    uint256 val;
+};
+
+template <>
+struct StackItem<true> {
+    uint256 val;
```

Comment:
We can move the `uint256 val` inside the `SymbolicStackItemPtr` object. Then the "symbolic" and swap and dup instructions are more efficient than the original.

Reply:
I tried this apporach, but it was a nightmare. I've reverted to having a "pure" value. However, we can keep the speed improvements, because now I only copy the pure value into the arena if a symbolic expression depends on it, e.g. a binop which has a non pure value in one of its children.